File: src/fio_ffi.erl
-module(fio_ffi).
-export([
    read_file/1,
    read_file_bits/1,
    write_file/2,
    write_file_bits/2,
    append_file/2,
    append_file_bits/2,
    delete_file/1,
    delete_directory/1,
    delete_directory_recursive/1,
    file_exists/1,
    file_info/1,
    link_info/1,
    is_directory/1,
    is_file/1,
    is_symlink/1,
    make_directory/1,
    make_directory_p/1,
    list_directory/1,
    copy_file/2,
    rename_file/2,
    create_symlink/2,
    create_hard_link/2,
    set_permissions/2,
    current_directory/0,
    get_tmp_dir/0,
    touch/1,
    read_link/1,
    checksum/2,
    write_file_atomic/2,
    write_file_bits_atomic/2,
    open_handle/2,
    close_handle/1,
    read_chunk/2,
    write_handle/2,
    write_handle_bits/2,
    seek_handle/2,
    tell_handle/1,
    unique_name/1
]).

%% --- Error mapping ---

map_error(Reason) ->
    case Reason of
        eacces -> {error, eacces};
        eagain -> {error, eagain};
        ebadf -> {error, ebadf};
        ebadmsg -> {error, ebadmsg};
        ebusy -> {error, ebusy};
        edeadlk -> {error, edeadlk};
        edquot -> {error, edquot};
        eexist -> {error, eexist};
        efault -> {error, efault};
        efbig -> {error, efbig};
        eintr -> {error, eintr};
        einval -> {error, einval};
        eio -> {error, eio};
        eisdir -> {error, eisdir};
        eloop -> {error, eloop};
        emfile -> {error, emfile};
        emlink -> {error, emlink};
        enametoolong -> {error, enametoolong};
        enfile -> {error, enfile};
        enodev -> {error, enodev};
        enoent -> {error, enoent};
        enomem -> {error, enomem};
        enospc -> {error, enospc};
        enosys -> {error, enosys};
        enotblk -> {error, enotblk};
        enotdir -> {error, enotdir};
        enotsup -> {error, enotsup};
        enxio -> {error, enxio};
        eoverflow -> {error, eoverflow};
        eperm -> {error, eperm};
        epipe -> {error, epipe};
        erange -> {error, erange};
        erofs -> {error, erofs};
        espipe -> {error, espipe};
        esrch -> {error, esrch};
        estale -> {error, estale};
        etxtbsy -> {error, etxtbsy};
        exdev -> {error, exdev};
        Other when is_atom(Other) -> {error, {unknown, atom_to_binary(Other, utf8)}};
        Other -> {error, {unknown, list_to_binary(io_lib:format("~p", [Other]))}}
    end.

%% --- File I/O ---

read_file(Path) ->
    case file:read_file(Path) of
        {ok, Bin} ->
            case unicode:characters_to_binary(Bin) of
                {error, _, _} -> {error, {not_utf8, Path}};
                {incomplete, _, _} -> {error, {not_utf8, Path}};
                Str when is_binary(Str) -> {ok, Str}
            end;
        {error, Reason} -> map_error(Reason)
    end.

read_file_bits(Path) ->
    case file:read_file(Path) of
        {ok, Bin} -> {ok, Bin};
        {error, Reason} -> map_error(Reason)
    end.

write_file(Path, Content) ->
    case file:write_file(Path, Content) of
        ok -> {ok, nil};
        {error, Reason} -> map_error(Reason)
    end.

write_file_bits(Path, Content) ->
    case file:write_file(Path, Content) of
        ok -> {ok, nil};
        {error, Reason} -> map_error(Reason)
    end.

%% Atomic write: write to a sibling temp file, then rename(2).
%% The temp lives in the same directory (same fs mount) so rename(2)
%% is guaranteed to be atomic on POSIX. The unique suffix prevents
%% collisions under concurrent writers.
write_file_atomic(Path, Content) ->
    Dir = filename:dirname(Path),
    Unique = integer_to_binary(erlang:unique_integer([positive, monotonic])),
    TmpPath = filename:join(Dir, <<".__fio_tmp_", Unique/binary>>),
    case file:write_file(TmpPath, Content) of
        ok ->
            case file:rename(TmpPath, Path) of
                ok ->
                    {ok, nil};
                {error, Reason} ->
                    _ = file:delete(TmpPath), % Best-effort cleanup upon rename failure
                    {error, {atomic_failed, <<"rename">>, atom_to_binary(Reason, utf8)}}
            end;
        {error, Reason} ->
            {error, {atomic_failed, <<"write_temp">>, atom_to_binary(Reason, utf8)}}
    end.

write_file_bits_atomic(Path, Content) ->
    write_file_atomic(Path, Content).

append_file(Path, Content) ->
    case file:write_file(Path, Content, [append]) of
        ok -> {ok, nil};
        {error, Reason} -> map_error(Reason)
    end.

append_file_bits(Path, Content) ->
    case file:write_file(Path, Content, [append]) of
        ok -> {ok, nil};
        {error, Reason} -> map_error(Reason)
    end.

%% --- Delete ---

delete_file(Path) ->
    case file:delete(Path) of
        ok -> {ok, nil};
        {error, Reason} -> map_error(Reason)
    end.

delete_directory(Path) ->
    case file:del_dir(Path) of
        ok -> {ok, nil};
        {error, Reason} -> map_error(Reason)
    end.

delete_directory_recursive(Path) ->
    %% Do not follow symlinks: if Path is a symlink, remove it rather than
    %% recursing into its target (prevents accidental deletion outside root).
    case file:read_link_info(Path, [{time, posix}]) of
        {ok, {file_info, _, symlink, _, _, _, _, _, _, _, _, _, _, _}} ->
            delete_file(Path);
        {ok, {file_info, _, directory, _, _, _, _, _, _, _, _, _, _, _}} ->
            case file:list_dir(Path) of
                {ok, Entries} ->
                    Results = lists:map(fun(Entry) ->
                        FullPath = filename:join(Path, Entry),
                        delete_directory_recursive(FullPath)
                    end, Entries),
                    case lists:filter(fun({error, _}) -> true; (_) -> false end, Results) of
                        [] -> delete_directory(Path);
                        [FirstError | _] -> FirstError
                    end;
                {error, Reason} -> map_error(Reason)
            end;
        {ok, _} ->
            %% Not a directory (regular file, device, etc.)
            delete_file(Path);
        {error, enoent} ->
            {ok, nil};
        {error, Reason} ->
            map_error(Reason)
    end.

%% --- File info ---

file_exists(Path) ->
    case file:read_file_info(Path) of
        {ok, _} -> true;
        {error, enoent} -> false;
        _ -> false
    end.

file_info(Path) ->
    case file:read_file_info(Path, [{time, posix}]) of
        {ok, {file_info, Size, _Type, _Access, Atime, Mtime, Ctime,
                         Mode, Nlinks, MajDev, MinDev, Inode, Uid, Gid}} ->
            Dev = (MajDev bsl 16) bor MinDev,
            {ok, {Size, Mode, Nlinks, Inode, Uid, Gid, Dev, Atime, Mtime, Ctime}};
        {error, Reason} -> map_error(Reason)
    end.

link_info(Path) ->
    case file:read_link_info(Path, [{time, posix}]) of
        {ok, {file_info, Size, _Type, _Access, Atime, Mtime, Ctime,
                         Mode, Nlinks, MajDev, MinDev, Inode, Uid, Gid}} ->
            Dev = (MajDev bsl 16) bor MinDev,
            {ok, {Size, Mode, Nlinks, Inode, Uid, Gid, Dev, Atime, Mtime, Ctime}};
        {error, Reason} -> map_error(Reason)
    end.

is_directory(Path) ->
    case file:read_file_info(Path, [{time, posix}]) of
        {ok, {file_info, _, directory, _, _, _, _, _, _, _, _, _, _, _}} -> {ok, true};
        {ok, _} -> {ok, false};
        {error, Reason} -> map_error(Reason)
    end.

is_file(Path) ->
    case file:read_file_info(Path, [{time, posix}]) of
        {ok, {file_info, _, regular, _, _, _, _, _, _, _, _, _, _, _}} -> {ok, true};
        {ok, _} -> {ok, false};
        {error, Reason} -> map_error(Reason)
    end.

is_symlink(Path) ->
    case file:read_link_info(Path, [{time, posix}]) of
        {ok, {file_info, _, symlink, _, _, _, _, _, _, _, _, _, _, _}} -> {ok, true};
        {ok, _} -> {ok, false};
        {error, Reason} -> map_error(Reason)
    end.

%% --- Directory operations ---

make_directory(Path) ->
    case file:make_dir(Path) of
        ok -> {ok, nil};
        {error, Reason} -> map_error(Reason)
    end.

make_directory_p(Path) ->
    case filelib:ensure_dir(filename:join(Path, "dummy")) of
        ok ->
            case file:make_dir(Path) of
                ok -> {ok, nil};
                {error, eexist} -> {ok, nil};
                {error, Reason} -> map_error(Reason)
            end;
        {error, Reason} -> map_error(Reason)
    end.

list_directory(Path) ->
    case file:list_dir(Path) of
        {ok, Entries} ->
            {ok, lists:map(fun(E) -> unicode:characters_to_binary(E) end, Entries)};
        {error, Reason} -> map_error(Reason)
    end.

%% --- Copy & Rename ---

copy_file(Src, Dest) ->
    case file:copy(Src, Dest) of
        {ok, _BytesCopied} -> {ok, nil};
        {error, Reason} -> map_error(Reason)
    end.

rename_file(Src, Dest) ->
    case file:rename(Src, Dest) of
        ok -> {ok, nil};
        {error, Reason} -> map_error(Reason)
    end.

%% --- Symlinks & Links ---

create_symlink(Target, Link) ->
    case file:make_symlink(Target, Link) of
        ok -> {ok, nil};
        {error, Reason} -> map_error(Reason)
    end.

create_hard_link(Target, Link) ->
    case file:make_link(Target, Link) of
        ok -> {ok, nil};
        {error, Reason} -> map_error(Reason)
    end.

%% --- Permissions ---

set_permissions(Path, Mode) ->
    case file:change_mode(Path, Mode) of
        ok -> {ok, nil};
        {error, Reason} -> map_error(Reason)
    end.

%% --- Utility ---

current_directory() ->
    case file:get_cwd() of
        {ok, Cwd} -> {ok, unicode:characters_to_binary(Cwd)};
        {error, Reason} -> map_error(Reason)
    end.

get_tmp_dir() ->
    Dir = case os:getenv("TMPDIR") of
        false ->
            case os:getenv("TEMP") of
                false ->
                    case os:getenv("TMP") of
                        false -> "/tmp";
                        Val -> Val
                    end;
                Val -> Val
            end;
        Val -> Val
    end,
    unicode:characters_to_binary(Dir).

unique_name(Prefix) ->
    Unique = integer_to_binary(erlang:unique_integer([positive, monotonic])),
    <<Prefix/binary, Unique/binary>>.

%% --- Touch & Link reading ---

touch(Path) ->
    case file:read_file_info(Path) of
        {ok, _} ->
            Now = calendar:local_time(),
            case file:change_time(Path, Now) of
                ok -> {ok, nil};
                {error, Reason} -> map_error(Reason)
            end;
        {error, enoent} ->
            case file:write_file(Path, <<>>) of
                ok -> {ok, nil};
                {error, Reason} -> map_error(Reason)
            end;
        {error, Reason} -> map_error(Reason)
    end.

read_link(Path) ->
    case file:read_link(Path) of
        {ok, Target} -> {ok, unicode:characters_to_binary(Target)};
        {error, Reason} -> map_error(Reason)
    end.

checksum(Path, Algo) ->
    case file:read_file(Path) of
        {ok, Bin} ->
            AlgoAtom = hash_algorithm(Algo),
            Hash = crypto:hash(AlgoAtom, Bin),
            {ok, bin_to_hex(Hash)};
        {error, Reason} -> map_error(Reason)
    end.

hash_algorithm(<<"sha256">>) -> sha256;
hash_algorithm(<<"sha512">>) -> sha512;
hash_algorithm(<<"md5">>)    -> md5;
hash_algorithm(_)            -> erlang:error(badarg).

bin_to_hex(Bin) ->
    list_to_binary([io_lib:format("~2.16.0b", [X]) || X <- binary_to_list(Bin)]).

%% --- File Handles ---
%%
%% open_handle returns the IoDevice pid returned by file:open/2.
%% It is carried as an opaque term through Gleam's RawHandle type.
%%
%% read_chunk returns {ok, {some, Data}} for a chunk or {ok, none} at EOF.
%% Gleam represents Some(x) as {some, x} and None as `none`.

open_handle(Path, Mode) ->
    Flags = case Mode of
        <<"r">> -> [read, binary];
        <<"w">> -> [write, binary];
        <<"a">> -> [append, binary];
        _       -> [read, binary]
    end,
    case file:open(Path, Flags) of
        {ok, Fd}        -> {ok, Fd};
        {error, Reason} -> map_error(Reason)
    end.

close_handle(Fd) ->
    case file:close(Fd) of
        ok              -> {ok, nil};
        {error, Reason} -> map_error(Reason)
    end.

read_chunk(Fd, Size) ->
    case file:read(Fd, Size) of
        {ok, Data}      -> {ok, {some, Data}};
        eof             -> {ok, none};
        {error, Reason} -> map_error(Reason)
    end.

write_handle(Fd, Content) ->
    case file:write(Fd, Content) of
        ok              -> {ok, nil};
        {error, Reason} -> map_error(Reason)
    end.

write_handle_bits(Fd, Content) ->
    case file:write(Fd, Content) of
        ok              -> {ok, nil};
        {error, Reason} -> map_error(Reason)
    end.

%% Move the file cursor to an absolute byte offset from the start of the file.
%% Uses file:position/2 which maps directly to lseek(2) on POSIX systems.
seek_handle(Fd, Position) ->
    case file:position(Fd, {bof, Position}) of
        {ok, _NewPos}   -> {ok, nil};
        {error, Reason} -> map_error(Reason)
    end.

%% Return the current byte offset of the file cursor.
tell_handle(Fd) ->
    case file:position(Fd, cur) of
        {ok, Pos}       -> {ok, Pos};
        {error, Reason} -> map_error(Reason)
    end.

File: src/fio_ffi.mjs

import * as fs from "node:fs";
import * as path from "node:path";
import * as os from "node:os";
import * as crypto from "node:crypto";
import { Ok, Error } from "./gleam.mjs";
import { toList, BitArray } from "../prelude.mjs";
import { Some, None } from "../gleam_stdlib/gleam/option.mjs";


function mapError(error) {
  const code = error.code || error.name || "UNKNOWN";
  switch (code) {
    case "EACCES": return new Error({ type: "eacces" });
    case "EAGAIN": return new Error({ type: "eagain" });
    case "EBADF": return new Error({ type: "ebadf" });
    case "EBUSY": return new Error({ type: "ebusy" });
    case "EEXIST": return new Error({ type: "eexist" });
    case "EFAULT": return new Error({ type: "efault" });
    case "EFBIG": return new Error({ type: "efbig" });
    case "EINTR": return new Error({ type: "eintr" });
    case "EINVAL": return new Error({ type: "einval" });
    case "EIO": return new Error({ type: "eio" });
    case "EISDIR": return new Error({ type: "eisdir" });
    case "ELOOP": return new Error({ type: "eloop" });
    case "EMFILE": return new Error({ type: "emfile" });
    case "EMLINK": return new Error({ type: "emlink" });
    case "ENAMETOOLONG": return new Error({ type: "enametoolong" });
    case "ENFILE": return new Error({ type: "enfile" });
    case "ENODEV": return new Error({ type: "enodev" });
    case "ENOENT": return new Error({ type: "enoent" });
    case "ENOMEM": return new Error({ type: "enomem" });
    case "ENOSPC": return new Error({ type: "enospc" });
    case "ENOSYS": return new Error({ type: "enosys" });
    case "ENOTDIR": return new Error({ type: "enotdir" });
    case "ENOTSUP": return new Error({ type: "enotsup" });
    case "ENOTEMPTY": return new Error({ type: "enotempty" });
    case "ENXIO": return new Error({ type: "enxio" });
    case "EOVERFLOW": return new Error({ type: "eoverflow" });
    case "EPERM": return new Error({ type: "eperm" });
    case "EPIPE": return new Error({ type: "epipe" });
    case "ERANGE": return new Error({ type: "erange" });
    case "EROFS": return new Error({ type: "erofs" });
    case "ESPIPE": return new Error({ type: "espipe" });
    case "ESRCH": return new Error({ type: "esrch" });
    case "ESTALE": return new Error({ type: "estale" });
    case "ETXTBSY": return new Error({ type: "etxtbsy" });
    case "EXDEV": return new Error({ type: "exdev" });
    default: return new Error({ type: "unknown", inner: String(error.message || code) });
  }
}


export function read_file(filePath) {
  try {
    const buf = fs.readFileSync(filePath);
    try {
      const decoder = new TextDecoder("utf-8", { fatal: true });
      return new Ok(decoder.decode(buf));
    } catch (_) {
      return new Error({ type: "not_utf8", path: filePath });
    }
  } catch (error) {
    return mapError(error);
  }
}

export function read_file_bits(filePath) {
  try {
    const buf = fs.readFileSync(filePath);
    return new Ok(toBitArray(buf));
  } catch (error) {
    return mapError(error);
  }
}

export function write_file(filePath, content) {
  try {
    fs.writeFileSync(filePath, content, "utf8");
    return new Ok(undefined);
  } catch (error) {
    return mapError(error);
  }
}

export function write_file_bits(filePath, bits) {
  try {
    fs.writeFileSync(filePath, fromBitArray(bits));
    return new Ok(undefined);
  } catch (error) {
    return mapError(error);
  }
}

export function append_file(filePath, content) {
  try {
    fs.appendFileSync(filePath, content, "utf8");
    return new Ok(undefined);
  } catch (error) {
    return mapError(error);
  }
}

export function append_file_bits(filePath, bits) {
  try {
    fs.appendFileSync(filePath, fromBitArray(bits));
    return new Ok(undefined);
  } catch (error) {
    return mapError(error);
  }
}


export function delete_file(filePath) {
  try {
    fs.unlinkSync(filePath);
    return new Ok(undefined);
  } catch (error) {
    return mapError(error);
  }
}

export function delete_directory(dirPath) {
  try {
    fs.rmdirSync(dirPath);
    return new Ok(undefined);
  } catch (error) {
    return mapError(error);
  }
}

export function delete_directory_recursive(dirPath) {
  try {
    fs.rmSync(dirPath, { recursive: true, force: true });
    return new Ok(undefined);
  } catch (error) {
    return mapError(error);
  }
}


export function file_exists(filePath) {
  return fs.existsSync(filePath);
}

export function file_info(filePath) {
  try {
    const stat = fs.statSync(filePath);
    return new Ok(statToFileInfo(stat));
  } catch (error) {
    return mapError(error);
  }
}

export function link_info(filePath) {
  try {
    const stat = fs.lstatSync(filePath);
    return new Ok(statToFileInfo(stat));
  } catch (error) {
    return mapError(error);
  }
}

function statToFileInfo(stat) {
  return [
    Number(stat.size),
    stat.mode,
    stat.nlink,
    Number(stat.ino),
    stat.uid,
    stat.gid,
    stat.dev,
    Math.floor(stat.atimeMs / 1000),
    Math.floor(stat.mtimeMs / 1000),
    Math.floor(stat.ctimeMs / 1000),
  ];
}

export function is_directory(filePath) {
  try {
    const stat = fs.statSync(filePath);
    return new Ok(stat.isDirectory());
  } catch (error) {
    return mapError(error);
  }
}

export function is_file(filePath) {
  try {
    const stat = fs.statSync(filePath);
    return new Ok(stat.isFile());
  } catch (error) {
    return mapError(error);
  }
}

export function is_symlink(filePath) {
  try {
    const stat = fs.lstatSync(filePath);
    return new Ok(stat.isSymbolicLink());
  } catch (error) {
    return mapError(error);
  }
}


export function make_directory(dirPath) {
  try {
    fs.mkdirSync(dirPath);
    return new Ok(undefined);
  } catch (error) {
    return mapError(error);
  }
}

export function make_directory_p(dirPath) {
  try {
    fs.mkdirSync(dirPath, { recursive: true });
    return new Ok(undefined);
  } catch (error) {
    return mapError(error);
  }
}

export function list_directory(dirPath) {
  try {
    const entries = fs.readdirSync(dirPath);
    return new Ok(toList(entries));
  } catch (error) {
    return mapError(error);
  }
}

// --- Copy & Rename ---

export function copy_file(src, dest) {
  try {
    fs.copyFileSync(src, dest);
    return new Ok(undefined);
  } catch (error) {
    return mapError(error);
  }
}

export function rename_file(src, dest) {
  try {
    fs.renameSync(src, dest);
    return new Ok(undefined);
  } catch (error) {
    return mapError(error);
  }
}

// --- Symlinks & Links ---

export function create_symlink(target, linkPath) {
  try {
    fs.symlinkSync(target, linkPath);
    return new Ok(undefined);
  } catch (error) {
    return mapError(error);
  }
}

export function create_hard_link(target, linkPath) {
  try {
    fs.linkSync(target, linkPath);
    return new Ok(undefined);
  } catch (error) {
    return mapError(error);
  }
}

// --- Permissions ---

export function set_permissions(filePath, mode) {
  try {
    fs.chmodSync(filePath, mode);
    return new Ok(undefined);
  } catch (error) {
    return mapError(error);
  }
}

// --- Utility ---

export function current_directory() {
  try {
    return new Ok(process.cwd());
  } catch (error) {
    return mapError(error);
  }
}

export function get_tmp_dir() {
  return os.tmpdir();
}

export function unique_name(prefix) {
  return prefix + Date.now() + Math.random().toString(36).slice(2);
}

// --- Touch & Link reading ---

export function touch(filePath) {
  try {
    try {
      const now = new Date();
      fs.utimesSync(filePath, now, now);
    } catch (e) {
      if (e.code === "ENOENT") {
        fs.writeFileSync(filePath, "");
      } else {
        throw e;
      }
    }
    return new Ok(undefined);
  } catch (error) {
    return mapError(error);
  }
}

export function read_link(filePath) {
  try {
    const target = fs.readlinkSync(filePath);
    return new Ok(target);
  } catch (error) {
    return mapError(error);
  }
}

export function checksum(filePath, algo) {
  try {
    const hash = crypto.createHash(algo);
    const buf = fs.readFileSync(filePath);
    hash.update(buf);
    return new Ok(hash.digest("hex"));
  } catch (error) {
    return mapError(error);
  }
}

// --- Atomic Write ---

// Writes `content` to a sibling temp file in the same directory, then
// calls fs.renameSync which is atomic on POSIX (same mount point).
// If anything fails the temp file is cleaned up before returning Error.
export function write_file_atomic(filePath, content) {
  const dir = path.dirname(filePath);
  const tmpPath = path.join(
    dir,
    `.__fio_tmp_${Date.now()}_${Math.random().toString(36).slice(2)}`
  );
  try {
    fs.writeFileSync(tmpPath, content, "utf8");
    try {
      fs.renameSync(tmpPath, filePath);
      return new Ok(undefined);
    } catch (renameErr) {
      try { fs.unlinkSync(tmpPath); } catch (_) { }
      return new Error({
        type: "atomic_failed",
        operation: "rename",
        reason: String(renameErr.message),
      });
    }
  } catch (writeErr) {
    return new Error({
      type: "atomic_failed",
      operation: "write_temp",
      reason: String(writeErr.message),
    });
  }
}

export function write_file_bits_atomic(filePath, bits) {
  const dir = path.dirname(filePath);
  const tmpPath = path.join(
    dir,
    `.__fio_tmp_${Date.now()}_${Math.random().toString(36).slice(2)}`
  );
  try {
    fs.writeFileSync(tmpPath, fromBitArray(bits));
    try {
      fs.renameSync(tmpPath, filePath);
      return new Ok(undefined);
    } catch (renameErr) {
      try { fs.unlinkSync(tmpPath); } catch (_) { }
      return new Error({
        type: "atomic_failed",
        operation: "rename",
        reason: String(renameErr.message),
      });
    }
  } catch (writeErr) {
    return new Error({
      type: "atomic_failed",
      operation: "write_temp",
      reason: String(writeErr.message),
    });
  }
}

// --- File Handles ---
//
// The JS handle is { fd, pos, isAppend } so we can:
//   - Track the cursor position explicitly for seek/tell/explicit reads.
//   - Detect append mode so writes use `null` (OS-forced EOF) rather than
//     an explicit position. Node.js ignores the 'a' flag when a numeric
//     position is passed to writeSync, so we must guard against that.
//
// read_chunk returns Ok(Some(BitArray)) for a chunk, Ok(None) on EOF.

export function open_handle(filePath, mode) {
  try {
    const fd = fs.openSync(filePath, mode);
    const isAppend = mode === "a";
    // For append mode seed `pos` at the current EOF so tell() is meaningful.
    let pos = 0;
    if (isAppend) {
      try { pos = fs.fstatSync(fd).size; } catch (_) { }
    }
    return new Ok({ fd, pos, isAppend });
  } catch (error) {
    return mapError(error);
  }
}

export function close_handle(handle) {
  try {
    fs.closeSync(handle.fd);
    return new Ok(undefined);
  } catch (error) {
    return mapError(error);
  }
}

export function read_chunk(handle, size) {
  try {
    const buf = Buffer.alloc(size);
    const bytesRead = fs.readSync(handle.fd, buf, 0, size, handle.pos);
    if (bytesRead === 0) {
      return new Ok(new None());
    }
    handle.pos += bytesRead;
    return new Ok(new Some(toBitArray(buf.slice(0, bytesRead))));
  } catch (error) {
    return mapError(error);
  }
}

export function write_handle(handle, content) {
  try {
    // In append mode always use null so the OS forces writes to EOF,
    // regardless of our tracked `pos`. Explicit position ignores 'a' flag.
    const position = handle.isAppend ? null : handle.pos;
    const bytesWritten = fs.writeSync(handle.fd, content, position, "utf8");
    handle.pos += bytesWritten;
    return new Ok(undefined);
  } catch (error) {
    return mapError(error);
  }
}

export function write_handle_bits(handle, bits) {
  try {
    const buf = fromBitArray(bits);
    const position = handle.isAppend ? null : handle.pos;
    const bytesWritten = fs.writeSync(handle.fd, buf, 0, buf.length, position);
    handle.pos += bytesWritten;
    return new Ok(undefined);
  } catch (error) {
    return mapError(error);
  }
}

export function seek_handle(handle, position) {
  // No OS call needed: update our tracked position.
  // In append mode seek affects read_chunk but writes still go to EOF.
  handle.pos = position;
  return new Ok(undefined);
}

export function tell_handle(handle) {
  return new Ok(handle.pos);
}

// --- Bitwise helpers ---

export function bitwise_and(a, b) {
  return a & b;
}

export function bitwise_shift_right(a, b) {
  return a >> b;
}

export function bitwise_shift_left(a, b) {
  return a << b;
}

// --- BitArray helpers ---

function toBitArray(buffer) {
  return new BitArray(new Uint8Array(buffer));
}

function fromBitArray(bitArray) {
  // Optimization: if the bit array is byte-aligned (offset 0 and integral bytes),
  // we can use the raw buffer directly. Buffer.from(Uint8Array) in Node.js is 
  // highly optimized.
  if (bitArray.bitOffset === 0 && bitArray.bitSize % 8 === 0) {
    return Buffer.from(bitArray.rawBuffer);
  }

  // Fallback for unaligned bit arrays: we must copy and shift bits.
  const buf = new Uint8Array(bitArray.byteSize);
  for (let i = 0; i < bitArray.byteSize; i++) {
    buf[i] = bitArray.byteAt(i);
  }
  return Buffer.from(buf);
}
